packages/api/src/functions/graphql.ts

```diff
@@ -40,11 +41,15 @@ export const createContextHandler = (
     // Get the authorization information from the request headers and request context.
     const type = getAuthProviderType(event)
     if (typeof type !== 'undefined') {
-      const authToken = await decodeAuthToken({ type, event, context })
+      const decoded = await decodeAuthToken({ type, event, context })
+      const token = await accessToken(event)
```

Comment:
I think this logic should be moved outside of this `graphql.ts` function, and into it's own function, maybe `getAuthenticationContext`

```js

const provider = authHeaders.getAuthProvider(event)
if (typeof provider === 'undefined') {
    return 
}

// the client is attempting to authenticate, kick-off the authentication mechanisms.
const {token }  = authHeaders.getAuthorization(event)

// we support decoding netlify, goTrue, and auth0 providers,
// other providers are "passed through" and decoding should be handled externally.
const decoded = await decodeAuthorizationToken({ token, provider })


return [decoded, { provider, token }]
```

Reply:
>  logic should be moved outside of this graphql.ts function

Yes, let's keep auth in auth and graph in graph -- so-to-speak.

Oh good, I was wondering if you'd be ok with:

```
return [decoded, { provider, token }]
```

Reply:
@peterp I've submitted the requested changes (`AuthorizationHeader`, getAuthorization in authHeaders, remove unneeded decoders, etc) and tested with Auth0.

Last part is to refactor the above in `graphql` for `getAuthenticationContext`.

Figuring out if I do a `decodeAuthorizationToken` as suggested how that changes the `AuthDecoder` interface .. or ... is `decodeAuthorizationToken` just in `getAuthenticationContext`. 

Because ...

```
      const decoded = await decodeAuthToken({ type, event, context })
      const authorization = getAuthorization(event)
      const token = authorization['token']
```

effectively fetches the token twice ... once while decoding and then a second time to send to `getCurrentUser` ... and no real need for that. I guess the decoder should just return:

`[decoded, { provider, token }]`

Will get to that later today.

Reply:
@peterp ok will get back to work on this today and implement the `getAuthenticationContext` as you request. That should be the last requested change for moment. I want to see if I don't have to call `getAuthorization` twice.

Reply:
@peterp I've implemented `getAuthenticationContext` in `AuthHeaders` and refactored `graphql`.

Some notes:

* I kept the parameter `type` instead of `provider`. I can change but wasn't sure that it would be confusing to still us `SupportedAuthTypes` or `SupportedAuthTypes` should become `SupportedAuthProviders`?
* getAuthorization still called twice (wanted to get it working first before changes downstream)
* This is a little ugly

```
    return [decoded, { type, token }]
  } else {
    return [null, { type: null, token: null }]
  }
```

Should we make a type for `decoded, { type, token }` and return that type? Or interface?

* I have a failing test still to fix:

`FAIL  src/__tests__/graphql.test.ts`

Something like

Reply:
The failing test was actually due to that somewhat ugly return

```
    return [decoded, { type, token }]
  } else {
    return [null, { type: undefined, token: null }]
```

My token was `null` but expected to be `undefined` in the logic check hence a null `currentUser` was always being added to the context.

```
@redwoodjs/api:   ● graphql createContextHandler › merges the context correctly
@redwoodjs/api:     expect(received).toEqual(expected) // deep equality
@redwoodjs/api:     - Expected  - 0
@redwoodjs/api:     + Received  + 1
@redwoodjs/api:       Object {
@redwoodjs/api:         "a": 1,
@redwoodjs/api:         "b": 2,
@redwoodjs/api:         "callbackWaitsForEmptyEventLoop": false,
@redwoodjs/api:     +   "currentUser": null,
```

@peterp ready for feedback and change requests